baseapp/abci.go

```diff
@@ -192,8 +192,21 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 			WithHeaderHash(req.Hash)
 	}
 
+	ctx := app.deliverState.ctx
+	if app.preBeginBlocker != nil {
+		app.preBeginBlocker(ctx, req)
+
+		// Manager skips this step if Block is non-nil since upgrade module is expected to set this params
+		// and consensus parameters should not be overwritten.
+		if cp := ctx.ConsensusParams(); cp == nil || cp.Block == nil {
```

Comment:
I still don't understand what does this check means.

Reply:
Same as paramsChanged

Reply:
How, aren't the handlers can't modify the ctx?

Reply:
but you called `ctx.ConsensusParams()`?

Reply:
`ctx.ConsensusParams()` will has params after `ctx = ctx.WithConsensusParams(cp)`, this is to ensure we only do once.

Reply:
I add a [noConsensusParams](https://github.com/crypto-org-chain/cosmos-sdk/pull/186/commits/cbb2d024a19da50008c80fe25a31aa0a63804580), see if it's better

Reply:
> this is to ensure we only do once.

why, what we really need is making sure it's reloaded if modified, right?

Reply:
> I add a [noConsensusParams](https://github.com/crypto-org-chain/cosmos-sdk/pull/186/commits/cbb2d024a19da50008c80fe25a31aa0a63804580), see if it's better

what we need here is to make sure it's reloaded if it's modified in `preBeginBlocker`, no matter if it's loaded before, if it's loaded before, just reload it, we just make sure the code after this will see the correct consensus parameters.
and we also need to modify `deliverState.ctx`, to make sure tx delivery and end blocker logic see correct consensus parameters as well.